File: CF-B/combinations.java
import java.math.BigInteger;
import java.util.HashMap;
import java.util.Scanner;

class Main
{
    public static void main(final String[] args)
    {
        HashMap<Integer, BigInteger> numbers = new HashMap<>();
        BigInteger p = new BigInteger("1");
        for (int i = 1; i <= 100; i++)
        {
            p = p.multiply(new BigInteger(String.valueOf(i)));
            numbers.put(i, p);
        }
        final Scanner scanner = new Scanner(System.in);
        Integer n, m;
        while ((n = scanner.nextInt()) != 0 && (m = scanner.nextInt()) != 0)
        {
            BigInteger div = numbers.get(n - m).multiply(numbers.get(m));
            BigInteger res = numbers.get(n).divide(div);
            System.out.println(n + " things taken " + m + " at a time is " + res + " exactly.");
        }
        scanner.close();
    }
}

File: CF-B/combinations.cpp
#include <iostream>
#include <set>
using namespace std;
using ull = unsigned long long;

int n, m;

int binomialCoeff(int a, int k)
{
    if (a < k) return 0;
    ull res = 1;
    if (k > a - k) k = a - k;
    for (int i = 0; i < k; i++)
    {
        res *= (a - i);
        res /= (i + 1);
    }
    return res;
}

int main() {
    while (cin >> n >> m && (n != 0 || m != 0)) {
        cout << n << " things taken " << m << " at a time is " << binomialCoeff(n, m) << " exactly." << endl;
    }
    return 0;
}

/*
100 things taken 6 at a time is 1192052400 exactly.
100 things taken 6 at a time is 1192052400 exactly.
*/